doc: update module customization hooks docs

GeoffreyBooth opened the pull request:
This PR builds on and follows up https://github.com/nodejs/node/pull/49261.

The aim here was to implement most of [this to-do list](https://github.com/nodejs/loaders/issues/154#issuecomment-1678068573). In particular:

- I hopefully removed all references to the “Loaders API” and similar nomenclature, in favor of “module customization hooks” or “hooks” or similar.

- I moved most of the “how to use this API” text out of the `module.register()` section and into a new section “Enabling” under the “Customization Hooks” heading.

- I removed all references to `--experimental-loader` except in `cli.md`, rewriting to recommend `--import` with `register()` instead. We’re considering removing the `--experimental-loader` flag, but for now at least we want to discourage its use. (We anticipate that most third-party libraries like `ts-node` will need code on the main thread, which `--import` with `register()` can enable but `--experimental-loader` cannot, so people should get familiar with the more powerful approach and we don’t need a secondary, less capable method.)

- I added CommonJS versions of all code samples that could support them (so anything running on the main thread, since the hooks module needs to export specific named functions).

- I classified the overall Module Customization Hooks API as stability level “1.1 - Active development” and the `resolve` and `load` hooks as “1.2 - Release candidate,” and `globalPreload` as “1.0 - Early development” (since I can’t mark it as deprecated, but its end is near whenever #49144 lands).

- I added some mentions that parts of this work for CommonJS modules and `require` as well.

- The CoffeeScript example seemed to have gotten out of date, so I rewrote it to how I think it should be now that CommonJS is supported.

- Miscellaneous cleanup.

What’s left for follow-up PRs:

- Remove the `globalPreload` section as part of #49144.

- Remove the mention of `--experimental-loader` from `cli.md`?

- Update `--experimental-loader` to print a warning that it may be removed in the future, and users should use `--import` with `register` instead; this warning could even generate example code that shows `--import` with a data URL that includes the value of what was passed to `--experimental-loader` (similar to how we have a fancy error message with suggested new code for unknown named exports from CommonJS modules).

Also fixes https://github.com/nodejs/node/issues/49282.

cc @nodejs/loaders @nodejs/modules @aduh95 @JakobJingleheimer

aduh95 commented on doc/api/cli.md:
```diff
@@ -43,9 +44,9 @@ Otherwise, the file is loaded using the CommonJS module loader. See
-command will only accept as input only files with `.js`, `.mjs`, or `.cjs`
-extensions; and with `.wasm` extensions when
+When loading [ECMAScript module loader][Modules loaders] loads the program
+entry point, the `node` command will only accept as input only files with
```
We probably want to remove one `only` from this sentence

aduh95 commented on doc/api/module.md:
```diff
@@ -269,80 +164,239 @@ changes:
+### Communication between main and hooks threads
+
+Module customization hooks run on a dedicated thread, separate from the main
+thread that runs application code. This means that global variables are unique
+to each thread, and message channels must be used to communicate between the
+threads.
```
```suggestion
thread that runs application code. This means mutating global variables won't affect
the other thread(s), and message channels must be used to communicate between the
threads.
```

aduh95 commented on doc/api/module.md:
```diff
@@ -269,80 +164,239 @@ changes:
+The [`register`][] method can be used to register a module that exports a set of
+hooks. The hooks are functions that are called by Node.js to customize the
+module resolution and loading process. The exported functions must have specific
+names and signatures, and they must be exported as named exports from an ES
+module file.
```
Should we start to deprecate using CJS on this loader thread? Currently exporting from a CJS module works just fine.

GeoffreyBooth replied:
We’ve never documented that hooks files could be CommonJS, so I think there’s nothing to deprecate? I’d rather not document it now because then it becomes a feature we need to continue to support (or deprecate) whereas in the current state we can remove that ability in the future (or more likely, inadvertently break that ability) and we haven’t created a bug.

Either way, I think this is a question beyond the scope of this PR. If we’re going to document this support then we should also add a test for it, which feels like a separate PR from this one.

aduh95 replied:
I think it's a different thing to have omissions in the docs (e.g. not mentioning that a CJS loader module works), and to have factually incorrect documentation (they must NOT be exported from an ES module file, it can be a CJS module, or a `data:`-URL module, or even HTTP with the ad-hoc flag). I suggest remove the last part of the sentence.
```suggestion
names and signatures, and they must be exported as named exports.
```

aduh95 commented on doc/api/module.md:
```diff
@@ -375,6 +429,30 @@ const result = register('/path-to-my-loader.js', {
+  assert.strictEqual(msg, 'increment: 2');
+});
+
+const result = register('/path-to-my-hooks.js', {
```
```suggestion
const result = register('./path-to-my-hooks.js', {
```

aduh95 commented on doc/api/module.md:
```diff
@@ -534,14 +607,15 @@ provided, all `require` calls from this module will be processed by the ESM
-The Node.js own `load` implementation, which is the value of `next` for the last
-loader in the `load` chain, returns `null` for `source` when `format` is
-`'commonjs'` for backward compatibility. Here is an example loader that would
+will not apply. (In other words, handled `require` calls will behave similarly
+to an `import` of a CommonJS module.) If `source` is undefined or `null`, it
```
> handled `require` calls will behave similarly to an `import` of a CommonJS module

Well not exactly, one is async, the other one is sync, that's quite a difference.

GeoffreyBooth commented on doc/api/module.md:
```diff
@@ -608,10 +608,11 @@ loader with registered `resolve` and `load` hooks; all `require.resolve` calls
-will be handled by the CommonJS module loader and `require`/`require.resolve`
-calls will not go through the registered hooks. This behavior for nullish
-`source` is temporary — in the future, nullish `source` will not be supported.
+to an `import` of a CommonJS module, other than `require` being sync.) If
```
No, the period goes inside the parentheses when the entire sentence is inside. The period would be outside only if the parenthetical were part of the sentence. https://blog.apastyle.org/apastyle/2013/03/punctuation-junction-periods-and-parentheses.html (see number 2).

aduh95 replied:
Wdyt about removing those parentheses?

GeoffreyBooth replied:
I guess that’s fine. If we do that we should maybe add a paragraph break after it, because it doesn’t really flow with the rest of the paragraph. It’s an aside:

> (In other words, handled `require` calls will behave similarly to an `import` of a CommonJS module, other than `require` being sync.)

GeoffreyBooth replied:
Want to suggest a specific change?

aduh95 commented on doc/api/module.md:
```diff
@@ -101,155 +101,850 @@ added: REPLACEME
+The value of `source` is ignored for type `'builtin'` because currently it is
+not possible to replace the value of a Node.js builtin (core) module.
+
+The value of `source` can be omitted for type `'commonjs'`. When a `source` is
+provided, all `require` calls from this module will be processed by the ESM
+loader with registered `resolve` and `load` hooks; all `require.resolve` calls
+from this module will be processed by the ESM loader with registered `resolve`
+hooks; `require.extensions` and monkey-patching on the CommonJS module loader
+will not apply. (In other words, handled `require` calls will behave similarly
+to an `import` of a CommonJS module, other than `require` being sync.) If
+`source` is undefined or `null`, it will be handled by the CommonJS module
+loader and `require`/`require.resolve` calls will not go through the registered
+hooks. This behavior for nullish `source` is temporary — in the future, nullish
+`source` will not be supported.
```
```suggestion
The value of `source` can be omitted for type `'commonjs'`:

- When a `source` is provided, all `require` calls from this module will be
  processed by the ESM loader with registered `resolve` and `load` hooks; all
  `require.resolve` calls from this module will be processed by the ESM loader
  with registered `resolve` hooks; only a subset of the CommonJS API will be
  available (e.g. no `require.extensions`, no `require.cache`, no
  `require.resolve.paths`) and monkey-patching on the CommonJS module loader
  will not apply.
- If `source` is undefined or `null`, it will be handled by the CommonJS module
  loader and `require`/`require.resolve` calls will not go through the
  registered hooks. This behavior for nullish `source` is temporary — in the
  future, nullish `source` will not be supported.
```

aduh95 commented on doc/api/module.md:
````diff
@@ -101,155 +101,852 @@ added: REPLACEME
+URL to `--import`:
+
+```bash
+node --import 'data:text/javascript,import { register } from "node:module"; register("http-to-https", import.meta.url);' ./my-app.js
````
```suggestion
node --import 'data:text/javascript,import { register } from "node:module"; import { fileToPathURL } from "node:url"; register("http-to-https", fileToPathURL('./'));' ./my-app.js
```

aduh95 commented on doc/api/module.md:
````diff
@@ -101,155 +101,852 @@ added: REPLACEME
+```
+
+With the preceding hooks module, running
+`node --import 'data:text/javascript,import { register } from "node:module"; register("./https-hooks.mjs", import.meta.url);' ./main.mjs`
````
```suggestion
`node --import 'data:text/javascript,import { register } from "node:module"; import { fileToPathURL } from "node:url"; register(fileToPathURL("./https-hooks.mjs"));' ./main.js`
```

aduh95 commented on doc/api/module.md:
````diff
@@ -101,155 +101,852 @@ added: REPLACEME
+console.log('some module!');
+```
+
+Running `node --import 'data:text/javascript,import { register } from "node:module"; register("./import-map-loader.js", import.meta.url);' ./import-map-loader.js main.js`
````
```suggestion
Running `node --import 'data:text/javascript,import { register } from "node:module"; import { fileToPathURL } from "node:url"; register(fileToPathURL("./import-map-loader.mjs"));' ./main.js`
```

aduh95 commented on doc/api/module.md:
```diff
@@ -803,7 +803,7 @@ console.log(VERSION);
 
 With the preceding hooks module, running
-`node --import 'data:text/javascript,import { register } from "node:module"; register("./https-hooks.mjs", import.meta.url);' ./main.mjs`
+`node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./https-hooks.mjs").href);' ./main.mjs`
```
```suggestion
`node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./https-hooks.mjs"));' ./main.mjs`
```

aduh95 commented on doc/api/module.md:
```diff
@@ -895,7 +895,7 @@ export scream = (str) -> str.toUpperCase()
 
 With the preceding hooks module, running
-`node --import 'data:text/javascript,import { register } from "node:module"; register("./coffeescript-hooks.mjs", import.meta.url);' main.coffee`
+`node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./coffeescript-hooks.mjs").href);' ./main.coffee`
```
```suggestion
`node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./coffeescript-hooks.mjs"));' ./main.coffee`
```

aduh95 commented on doc/api/module.md:
````diff
@@ -944,7 +944,7 @@ import 'a-module';
 ```
 
-Running `node --import 'data:text/javascript,import { register } from "node:module"; register("./import-map-loader.js", import.meta.url);' ./import-map-loader.js main.js`
+Running `node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./import-map-hooks.js").href);' ./import-map-loader.js main.js`
````
```suggestion
Running `node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./import-map-hooks.js"));' main.js`
```

aduh95 commented:
Maybe we could address https://github.com/nodejs/node/issues/49282 as part of this PR?

GeoffreyBooth commented:
> Maybe we could address #49282 as part of this PR?

How, exactly? The examples already show the functions’ signatures as being async.

aduh95 commented on doc/api/module.md:
````diff
@@ -269,65 +164,224 @@ changes:
+node --import some-package/register ./my-app.js
+```
+
+Where `some-package` has an `"exports"` field defining the `/register` export to
+map to a file that calls `register()`, like this `register-hooks.js` example.
````
I think we should either link to the `"exports"` documentation, or not mention it at all considering folks already know – and realistically, they will just copy the command from `some-package` README, so they don't have to care about exports map or this kind of stuff.

GeoffreyBooth replied:
I'm fine with not mentioning it, but will people be confused?

aduh95 commented on doc/api/module.md:
```diff
@@ -269,65 +164,224 @@ changes:
+`second.mjs` define a `resolve` hook, both will be called, in the order they
+were registered. The same applies to all the other hooks.
+
+### Communication between main and hooks threads
```
Should we keep calling it the loader thread?

GeoffreyBooth replied:
I don't think so. It's the thread where the hooks run. One term is better than two.

JakobJingleheimer replied:
Orchestrator/orchestration threads?

GeoffreyBooth replied:
Customization thread?

I wasn't assuming we'd need a thread like this (or use this one) for other systems, but who knows?

aduh95 commented on doc/api/module.md:
```diff
@@ -338,11 +392,10 @@ either:
 * a `Promise` resolving to one of the aforementioned values.
 
-Loader code:
+Hooks code:
```
```suggestion
Module customization code:
```

aduh95 commented on doc/api/module.md:
```diff
@@ -375,6 +428,30 @@ const result = register('/path-to-my-loader.js', {
+  transferList: [port2],
+});
+
+assert.strictEqual(result, 'ok');
```
I'm not sure we want to keep documenting that, it's not unlikely that we will get rid of this, or at least that it's going to change shape. I guess it's fine to keep it until we actually change it, but folks hould not rely on it.

GeoffreyBooth replied:
Yeah I don't know, I didn't change the example (just added the cjs version). Do you have a suggestion?

JakobJingleheimer replied:
I think the `'ok'`s are also a little confusing because some assertion libs have things like "expect().to.be.ok".

I would remove `initialize`'s return, and remove this assert.

GeoffreyBooth replied:
My goals in this PR were just to implement the [to-do list](https://github.com/nodejs/loaders/issues/154#issuecomment-1678068573); I’m fine with making additional improvements, but I don’t want this PR to drag on for days or weeks because we’re reopening _all_ the text on this page. My PR didn’t change this example; if you want to suggest an improvement, fine, I’ll probably accept it; but otherwise, I think improving this example (and other examples that I haven’t changed) can follow in a new PR.

JakobJingleheimer replied:
Ah, sorry, all I see is red and green.

GeoffreyBooth commented:
Per https://github.com/nodejs/node/issues/49282#issuecomment-1689071368 I think this PR addresses the concerns raised in https://github.com/nodejs/node/issues/49282.

JakobJingleheimer commented on doc/api/module.md:
````diff
@@ -269,65 +164,225 @@ changes:
+```
+
+Where `some-package` has an [`"exports"`][] field defining the `/register`
+export to map to a file that calls `register()`, like this `register-hooks.js`
+example.
````
nit: "this" is like the example should immediately follow, whereas "the following … example" is more like there "will be an example soon"

```suggestion
export to map to a file that calls `register()`, like the following `register-hooks.js`
example.
```

JakobJingleheimer commented on doc/api/module.md:
````diff
@@ -269,65 +164,225 @@ changes:
+`register` can be called from the entry point, but dynamic `import()` must be
+used for any code that should be run after the hooks are registered:
+
+```mjs
+import { register } from 'node:module';
````
```suggestion
```mjs
// register-hooks.mjs

import { register } from 'node:module';
```

GeoffreyBooth replied:
I don’t think the filenames are relevant, we’re not referring to this code block from an example command to be run.

JakobJingleheimer commented on doc/api/module.md:
````diff
@@ -269,65 +164,225 @@ changes:
+await import('./my-app.js');
+```
+
+```cjs
+const { register } = require('node:module');
````
```suggestion
```cjs
// register-hooks.cjs

const { register } = require('node:module');
```

JakobJingleheimer commented on doc/api/module.md:
````diff
@@ -269,65 +164,225 @@ changes:
+import('./my-app.js');
+```
+
+In this example, we are registering the `http-to-https` hooks, but they will
````
this paragraph is between 2 examples, so the one to which it applies is ambiguous

```suggestion
In the above example, we are registering the `http-to-https` hooks, but they will
```

GeoffreyBooth replied:
There was some accessibility guide that said we should use “previous” and “next” instead of “above” and “below”, I forget why. We follow the Microsoft style guide, it’s in there somewhere.

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+In this example, we are registering the `http-to-https` hooks, but they will
+only be available for subsequently imported modules—in this case, `my-app.js`
+and anything it references via `import` or `require`. If the
+`import('./my-app.js')` had instead been a static `import './my-app.js'`, _the
+app would already have been loaded_ before the `http-to-https` hooks were
```
less is more :)

```suggestion
`import('./my-app.js')` had instead been a static `import './my-app.js'`, the
app would have _already_ been loaded **before** the `http-to-https` hooks were
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+and anything it references via `import` or `require`. If the
+`import('./my-app.js')` had instead been a static `import './my-app.js'`, _the
+app would already have been loaded_ before the `http-to-https` hooks were
+registered. This is part of the design of ES modules, where static imports are
```
```suggestion
registered. This due to the ES modules specification, where static imports are
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+`import('./my-app.js')` had instead been a static `import './my-app.js'`, _the
+app would already have been loaded_ before the `http-to-https` hooks were
+registered. This is part of the design of ES modules, where static imports are
+evaluated from the leaves of the tree first back to the trunk. There can be
```
```suggestion
evaluated from the leaves of the tree first, then back to the trunk. There can be
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
-When hooks are used they apply to each subsequent loader, the entry point, and
-all `import` calls. They won't apply to `require` calls; those still follow
-[CommonJS][] rules.
+The `my-app.js` file can also be CommonJS. Customization hooks will run for any
```
```suggestion
`my-app.js` can also be CommonJS. Customization hooks will run for any
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+// entrypoint.mjs
+import { register } from 'node:module';
+
+register(new URL('./first.mjs', import.meta.url));
+register('./second.mjs', import.meta.url);
```
Is there any difference here, using `URL()` vs not? I believe no, so without an explanation that they're the same, it's a hanging-chad.

```suggestion
register('./first.mjs', import.meta.url);
register('./second.mjs', import.meta.url);
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+const { pathToFileURL } = require('node:url');
+
+const parentURL = pathToFileURL(__filename);
+register(new URL('./first.mjs', parentURL));
+register('./second.mjs', parentURL);
```
```suggestion
register('./first.mjs', parentURL);
register('./second.mjs', parentURL);
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+### Communication between main and hooks threads
+
+Module customization hooks run on a dedicated thread, separate from the main
+thread that runs application code. This means mutating global variables won't
```
nit

```suggestion
thread runing application code. This means mutating global variables won't
```

GeoffreyBooth replied:
I’m ambivalent about this one but I think the old version is better. “Running” implies that something is continuously happening, but in reality there are periods of idleness.

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+import { register } from 'node:module';
+import { MessageChannel } from 'node:worker_threads';
+
+// This example showcases how a message channel can be used to
```
nit ("showcase" seems a bit much)

```suggestion
// This example demonstrates how a message channel can be used to
```

GeoffreyBooth replied:
This example is an extravaganza!

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+import { MessageChannel } from 'node:worker_threads';
+
+// This example showcases how a message channel can be used to
+// communicate to the hooks, by sending `port2` to the hooks.
```
communication is bi-directional ("to" is uni-directional)

```suggestion
// communicate with the hooks, by sending `port2` to the hooks.
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
+const { MessageChannel } = require('node:worker_threads');
+
+// This example showcases how a message channel can be used to
+// communicate to the hooks, by sending `port2` to the hooks.
```
```suggestion
// communicate with the hooks, by sending `port2` to the hooks.
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
 
 ### Hooks
 
+The [`register`][] method can be used to register a module that exports a set of
```
```suggestion
The [`register`][] method can be used to register a module exporting a set of
```

GeoffreyBooth replied:
Same as the similar other example, this reads as less clear to me because “exporting” could be modifying multiple things whereas “that exports” can only refer to “a module.”

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -269,65 +164,225 @@ changes:
-different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
-terminated by the main thread at any time, so do not depend on asynchronous
-operations (like `console.log`) to complete.
+Hooks are run in a separate thread, isolated from the main thread where
+application code runs. That means it is a different [realm][]. The hooks thread
```
The first sentence here is a verbatim copy of the intro sentence of this section. Could we remove it here, move the second sentence ("That means …") up to the intro, and move the third into a callout/caveat?

GeoffreyBooth replied:
> The first sentence here is a verbatim copy of the intro sentence of this section.

It is? This used to be the introduction of the section, but it’s farther down now.

JakobJingleheimer replied:
😳 I can't find it now. Sorry, I don't know what I was looking at.

JakobJingleheimer commented on doc/api/module.md:
````diff
@@ -338,11 +393,10 @@ either:
 ```mjs
-// In the below example this file is referenced as
-// '/path-to-my-loader.js'
+// In the below example this file is referenced as `/path-to-my-hooks.js`.
````
KISS

```suggestion
// path-to-my-hooks.js
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -529,19 +602,23 @@ The final value of `format` must be one of the following:
-The Node.js own `load` implementation, which is the value of `next` for the last
-loader in the `load` chain, returns `null` for `source` when `format` is
-`'commonjs'` for backward compatibility. Here is an example loader that would
+The value of `source` can be omitted for type `'commonjs'`:
```
This sentence is confusing. It seems like it's about to explain the effects of omitting `source`, but it goes on to explain omitting vs not omitting.

```suggestion
Omitting vs providing a `source` for `'commonjs'` has very different effects:
```

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -529,19 +602,23 @@ The final value of `format` must be one of the following:
-`'commonjs'` for backward compatibility. Here is an example loader that would
+The value of `source` can be omitted for type `'commonjs'`:
+
+* When a `source` is provided, all `require` calls from this module will be
```
```suggestion
* When provided, all `require` calls from this module will be
```

GeoffreyBooth replied:
I think it’s clearer before, that “provided” refers to `source` and not “all `require` calls”.

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -529,19 +602,23 @@ The final value of `format` must be one of the following:
+  available (e.g. no `require.extensions`, no `require.cache`, no
+  `require.resolve.paths`) and monkey-patching on the CommonJS module loader
+  will not apply.
+* If `source` is undefined or `null`, it will be handled by the CommonJS module
```
```suggestion
* When omitted (`source`'s value is undefined or `null`), it will be handled by the CommonJS module
```

JakobJingleheimer replied:
nit: I would move this case first because

* this was the previous behaviour
* alphabetical order (o…p)

GeoffreyBooth replied:
Perhaps, but the omitting behavior is weird and shouldn’t be encouraged. Generally we want hooks to support CommonJS, so we should put the behavior we want first to imply that it’s the default.

GeoffreyBooth commented on doc/api/cli.md:
```diff
@@ -43,9 +44,9 @@ Otherwise, the file is loaded using the CommonJS module loader. See
-When loading [ECMAScript module loader][] loads the program entry point, the `node`
-command will only accept as input only files with `.js`, `.mjs`, or `.cjs`
-extensions; and with `.wasm` extensions when
+When loading [ECMAScript module loader][Modules loaders] loads the program entry
+point, the `node` command will accept as input only files with `.js`, `.mjs`, or
```
What is the “module loader” called by other runtimes? Also that rename can happen in its own PR, it probably involves changes across many docs files.

JakobJingleheimer replied:
Idk what other runtimes call it.

Sure, we can handle the name change in a different PR. You were changing other names, so it seemed appropriate, but not necessarily together.

GeoffreyBooth commented on doc/api/cli.md:
```diff
@@ -550,7 +551,11 @@ changes:
 -->
 
-Specify the `module` of a custom experimental [ECMAScript module loader][].
+> This flag is discouraged and may be removed in a future version of Node.js.
+> Please use
+> [`--import` with `register()`][module customization hooks: enabling] instead.
```
This doesn’t feel like an improvement. We don’t need to explain why it’s discouraged. It may be “less robust” but it’s shorter than `--import` with a `data:` URL, so there are tradeoffs. We don’t need to get into it in the docs.

GeoffreyBooth replied:
Perhaps, but I don’t think your suggestion really answers the “why” question, and to do so properly would require more space than is worth. I think the documentation just needs to explain how to use the APIs, not get into digressions about why decisions were made or current designs are better than previous ones. If someone truly wants to know then they’ll search for it and find GitHub issues explaining it.

GeoffreyBooth commented on doc/api/module.md:
```diff
@@ -102,114 +102,7 @@ added: REPLACEME
 * Returns: {any} returns whatever was returned by the `initialize` hook.
 
 Register a module that exports [hooks][] that customize Node.js module
```
Saying “exporting” here is unclear if you’re modifying “module” or “register a module”; as in, the “exporting” could be happening as a result of the registration. I think the existing text is clearer.

JakobJingleheimer reviewed (approved):
I'll handle my comments in a subsequent PR. looks fine to me

GeoffreyBooth commented on doc/api/module.md:
````diff
@@ -269,65 +164,225 @@ changes:
+imported from `node:module`, which you can run before your application code by
+using the `--import` flag:
+
+```bash
````
We seem to use both in our docs, `bash` more so than `console`.

JakobJingleheimer replied:
Then we have many places that are wrong 🤷‍♂️ In this case correctness is better than consistently wrong, no?

Bash is for a shell file (like `foo.sh`)

Console is for a CLI prompt: https://stackoverflow.com/a/49004070/758177

aduh95 replied:
Well that's `bash`, not `console`. `console` is when you see console output, `bash` is for shell script that you can copy paste to your terminal (or bash script)

aduh95 replied:
From the examples you just gave, wouldn't you agree that `bash` is more suited than `console` here?

JakobJingleheimer replied:
We _happen_ to not see a difference, but if the CLI snippet changes, it could incorrectly/needlessly break, and the editor may not know why.

I'm not saying "this absolutely must be fixed"; I'm saying it's wrong and easily fixed to avoid easily avoidable problems in future.

aduh95 replied:
We'll have to agree to disagree here, I would feel strongly that `console` is wrong here, the code snippet _is_ a `bash` command. I've spent quite some time on https://github.com/nodejs/node/commit/260092eca316bae6036e52b2cdc73c4c4f8a9da6 to make sure the doc were consistent on that matter.

GeoffreyBooth replied:
I’m going to mark this as resolved and you can open a new PR for this, perhaps to make it consistent throughout the codebase and enforced via lint rules. The existing examples used/use `bash` so I’m going to leave it that way for now.

GeoffreyBooth commented on doc/api/cli.md:
```diff
@@ -43,9 +44,9 @@ Otherwise, the file is loaded using the CommonJS module loader. See
-extensions; and with `.wasm` extensions when
+When loading [ECMAScript module loader][Modules loaders] loads the program entry
+point, the `node` command will accept as input only files with `.js`, `.mjs`, or
+`.cjs` extensions; and with `.wasm` extensions when
```
“the” extensions? I think the previous is clearer.

JakobJingleheimer replied:
Oh, yes. extensions → extension

The current separates them into different sentences and the first already contains "extensions".

Nbd though

aduh95 commented on doc/api/module.md:
```diff
@@ -417,21 +494,21 @@ changes:
+optionally return a format (such as `'module'`) as a hint to the `load` hook. If
+a format is specified, the `load` hook is ultimately responsible for providing
+the final `format` value (and it is free to ignore the hint provided by
+`resolve`); if `resolve` provides a `format`, a custom `load` hook is required
+even if only to pass the value to the Node.js default `load` hook.
```
While we're editing this, we should consider rephrase that, I don't understand what it means.
